Cannot extract contents of Renjin 3.5 beta76 archive

# Environment

```
$ gradle --version

------------------------------------------------------------
Gradle 7.5.1
------------------------------------------------------------

Build time:   2022-08-05 21:17:56 UTC
Revision:     d1daa0cbf1a0103000b71484e1dbfe096e095918

Kotlin:       1.6.21
Groovy:       3.0.10
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          18.0.2 (BellSoft 18.0.2+10)
OS:           Linux 5.18.10-arch1-1 amd64

$ java -version
openjdk version "18.0.2" 2022-07-19
OpenJDK Runtime Environment (build 18.0.2+10)
OpenJDK 64-Bit Server VM (build 18.0.2+10, mixed mode, sharing)
```

# Replicate

``` bash
mkdir -p $HOME/dev/fail
cd $HOME/dev/fail
cat << EOF > build.gradle
plugins { id 'java' }

repositories {
  mavenCentral()
  maven { url "https://nexus.bedatadriven.com/content/groups/public" }
}

dependencies { implementation 'org.renjin:renjin-script-engine:3.5-beta76' }

jar {
  duplicatesStrategy = DuplicatesStrategy.EXCLUDE

  from {
    (configurations.runtimeClasspath.findAll { !it.path.endsWith(".pom") })
    .collect {
      it.isDirectory() ? it : zipTree(it)
    }
  }
}
EOF
gradle jar
```

# Expected

No error.

# Actual

```
Configuration on demand is an incubating feature.
> Task :jar FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jar'.
> Could not expand ZIP '/home/username/.gradle/caches/modules-2/files-2.1/org.renjin/renjin-core/3.5-beta76/52ccd406ab3306df0d37929939d1e28b457057c7/renjin-core-3.5-beta76.jar'.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
1 actionable task: 1 executed
```

# Regression

Is this a regression? The same `gradle.build` file works with Gradle 7.3.

Both `jar` and `zip` commands can open the file successfully.

``` bash
$ zip --test /home/username/.gradle/caches/modules-2/files-2.1/org.renjin/renjin-core/3.5-beta76/52ccd406ab3306df0d37929939d1e28b457057c7/renjin-core-3.5-beta76.jar 
test of /home/username/.gradle/caches/modules-2/files-2.1/org.renjin/renjin-core/3.5-beta76/52ccd406ab3306df0d37929939d1e28b457057c7/renjin-core-3.5-beta76.jar OK
$ jar -tvf /home/username/.gradle/caches/modules-2/files-2.1/org.renjin/renjin-core/3.5-beta76/52ccd406ab3306df0d37929939d1e28b457057c7/renjin-core-3.5-beta76.jar | wc -l
2253
```

# Stacktrace

```
Caused by: org.gradle.api.GradleException: Could not expand ZIP '/home/username/.gradle/caches/modules-2/files-2.1/org.renjin/renjin-core/3.5-beta76/52ccd406ab3306df0d37929939d1e28b457057c7/renjin-core-3.5-beta76.jar'.
        at org.gradle.api.internal.file.archive.ZipFileTree.visit(ZipFileTree.java:108)
        at org.gradle.api.internal.file.collections.FilteredMinimalFileTree.visit(FilteredMinimalFileTree.java:89)
        at org.gradle.api.internal.file.collections.FileTreeAdapter.visit(FileTreeAdapter.java:96)
```

Confirmed that it works in Gradle `7.3.3`, fails in more recent versions.

Root cause:
```
Caused by: java.lang.IllegalArgumentException: 'org/renjin/base/..getNamespace.RData' is not a safe zip entry name.
        at org.gradle.util.internal.ZipSlip.safeZipEntryName(ZipSlip.java:37)
        at org.gradle.api.internal.file.archive.ZipFileTree$DetailsImpl.safeEntryName(ZipFileTree.java:172)
        at org.gradle.api.internal.file.archive.ZipFileTree$DetailsImpl.getRelativePath(ZipFileTree.java:201)
        at org.gradle.api.tasks.util.internal.CachingPatternSpecFactory$CachingSpec.isSatisfiedBy(CachingPatternSpecFactory.java:72)
        at org.gradle.api.tasks.util.internal.CachingPatternSpecFactory$CachingSpec.isSatisfiedBy(CachingPatternSpecFactory.java:61)
        at org.gradle.api.specs.NotSpec.isSatisfiedBy(NotSpec.java:36)
        at org.gradle.api.internal.file.collections.FilteredMinimalFileTree$2.visitFile(FilteredMinimalFileTree.java:99)
        at org.gradle.api.internal.file.archive.ZipFileTree.visit(ZipFileTree.java:104)
        ... 136 more
```

--- a/subprojects/wrapper-shared/src/main/java/org/gradle/util/internal/ZipSlip.java
+++ b/subprojects/wrapper-shared/src/main/java/org/gradle/util/internal/ZipSlip.java
@@ -43,8 +43,19 @@ public class ZipSlip {
         return name.isEmpty()
             || name.startsWith("/")
             || name.startsWith("\\")
-            || name.contains("..")
+            || containsDirectoryNavigation(name)
             || (name.contains(":") && isWindows());
+    }
+
+    private static boolean containsDirectoryNavigation(String name) {
+        if (!name.contains("..")) {
+            return false;
+        }
+        // We have a .. but if not before a file separator or at the end, it is OK
+        return name.endsWith("\\..")
+            || name.contains("..\\")
+            || name.endsWith("/..")
+            || name.contains("../");
     }
 
     private static boolean isWindows() {

--- a/subprojects/wrapper-shared/src/test/groovy/org/gradle/util/internal/ZipSlipTest.groovy
+++ b/subprojects/wrapper-shared/src/test/groovy/org/gradle/util/internal/ZipSlipTest.groovy
@@ -36,14 +36,18 @@ class ZipSlipTest extends Specification {
         !isUnsafeZipEntryName(safePath)
 
         where:
-        unsafePath | safePath
-        "/"        | "foo/"
-        "\\"       | "foo\\"
-        "/foo"     | "foo"
-        "\\foo"    | "foo"
-        "foo/.."   | "foo/bar"
-        "foo\\.."  | "foo\\bar"
-        "C:/foo"   | "foo"
+        unsafePath     | safePath
+        "/"            | "foo/"
+        "\\"           | "foo\\"
+        "/foo"         | "foo"
+        "\\foo"        | "foo"
+        "foo/.."       | "foo/bar"
+        "foo\\.."      | "foo\\bar"
+        "C:/foo"       | "foo"
+        "../foo"       | "..foo"
+        "..\\foo"      | "..foo"
+        "foo/../bar"   | "foo/..bar"
+        "foo\\..\\bar" | "foo\\..bar"
     }
 
     private static boolean isWindows() {

--- a/subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy
+++ b/subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperGenerationIntegrationTest.groovy
@@ -54,7 +54,7 @@ class WrapperGenerationIntegrationTest extends AbstractIntegrationSpec {
 
         then:
         // wrapper needs to be small. Let's check it's smaller than some arbitrary 'small' limit
-        file("gradle/wrapper/gradle-wrapper.jar").length() < 60 * 1024
+        file("gradle/wrapper/gradle-wrapper.jar").length() < 61 * 1024
     }
 
     def "generated wrapper scripts for given version from command-line"() {
@@ -71,7 +71,7 @@ class WrapperGenerationIntegrationTest extends AbstractIntegrationSpec {
         executer.inDirectory(file("second")).withTasks("wrapper").run()
 
         then: "the checksum should be constant (unless there are code changes)"
-        Hashing.sha256().hashFile(file("first/gradle/wrapper/gradle-wrapper.jar")) == HashCode.fromString("53018fb29b1af30cb2776fd1ddfede2d1a60bb541d00750cbbb0e40e7c61226b")
+        Hashing.sha256().hashFile(file("first/gradle/wrapper/gradle-wrapper.jar")) == HashCode.fromString("d2d670ead532ad47bf5bacae8aa8d8fb9e6b7c2aa992524931a416f121db1e63")
 
         and:
         file("first/gradle/wrapper/gradle-wrapper.jar").md5Hash == file("second/gradle/wrapper/gradle-wrapper.jar").md5Hash

